### PR TITLE
[SPARK-10102][Streaming]Fix a race condition that startReceiver may happen before setting trackerState to Started

### PR DESCRIPTION
**Pull request by zsxwing**
Test failure: https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-Maven-with-YARN/HADOOP_PROFILE=hadoop-2.4,label=spark-test/3305/testReport/junit/org.apache.spark.streaming/StreamingContextSuite/stop_gracefully/

There is a race condition that setting `trackerState` to `Started` could happen after calling `startReceiver`. Then `startReceiver` won't start the receivers because it uses `! isTrackerStarted` to check if ReceiverTracker is stopping or stopped. But actually, `trackerState` is `Initialized` and will be changed to `Started` soon.

Therefore, we should use `isTrackerStopping || isTrackerStopped`.

### CODE REVIEW

**Comment by zsxwing** [comment]
BTW, other usages of `isTrackerStarted` are right because they are in the `synchronized` body.

**Comment by zsxwing** [comment]
Fired a new JIRA since this is a bug in the main codes.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 1]
```diff
@@ -469,7 +469,7 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
      */
     private def startReceiver(receiver: Receiver[_], scheduledExecutors: Seq[String]): Unit = {
       val receiverId = receiver.streamId
-      if (!isTrackerStarted) {
+      if (isTrackerStopping || isTrackerStopped) {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: Can you make this a internal function? `def shouldStartReceiver: Boolean = !(isTrackerStopping || isTrackerStopped)`
and then use `if (!shouldStartReceiver) { ... }`

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: And add comments that its okay to start when the state = initialized / started

**Comment by tdas** [comment]
LGTM. Will merge when tests pass.

**Comment by tdas** [comment]
Merging this, thanks @zsxwing for finding this issue.

**Pull request closed without merging**